decompiler/IR2/FormExpressionAnalysis.cpp

```diff
@@ -3808,6 +3808,64 @@ FormElement* try_make_nonzero_logtest(Form* in, FormPool& pool) {
   return nullptr;
 }
 
+FormElement* try_make_logtest_cpad_macro(Form* in, FormPool& pool) {
+  /*
+(defmacro cpad-pressed (pad-idx)
+  `(-> *cpad-list* cpads ,pad-idx button0-rel 0)
+  )
+
+(defmacro cpad-hold (pad-idx)
+  `(-> *cpad-list* cpads ,pad-idx button0-abs 0)
+  )
+
+(defmacro cpad-pressed? (pad-idx &rest buttons)
+  `(logtest? (cpad-pressed ,pad-idx) (pad-buttons ,@buttons))
+  )
+
+(defmacro cpad-hold? (pad-idx &rest buttons)
+  `(logtest? (cpad-hold ,pad-idx) (pad-buttons ,@buttons))
+  )
+ */
+  auto cpad_matcher = Matcher::op(
+      GenericOpMatcher::fixed(FixedOperatorKind::LOGTEST),
+      {Matcher::deref(Matcher::symbol("*cpad-list*"), false,
+                      {DerefTokenMatcher::string("cpads"), DerefTokenMatcher::any_integer(0),
+                       DerefTokenMatcher::any_string(2), DerefTokenMatcher::integer(0)}),
+       Matcher::op_with_rest(GenericOpMatcher::func(Matcher::constant_token("pad-buttons")), {})});
+  auto mr = match(cpad_matcher, in);
+  if (mr.matched) {
+    enum { ABS, REL, NIL } t = NIL;
+    if (mr.maps.strings.at(2) == "button0-abs") {
+      t = ABS;
+    } else if (mr.maps.strings.at(2) == "button0-rel") {
+      t = REL;
+    }
+
+    if (t != NIL) {
+      auto jank = dynamic_cast<GenericElement*>(
+          pool.alloc_element<GenericElement>(
+                  GenericOperator::make_fixed(FixedOperatorKind::LOGTEST), in)
+              ->elts()
+              .at(0)
+              ->at(0));
+      auto buttons_form = jank->elts().at(1);
+      std::vector<Form*> v = {
+          pool.form<SimpleAtomElement>(SimpleAtom::make_int_constant(mr.maps.ints.at(0)))};
+      int i = 2;
+      buttons_form->apply_form([&](Form* form) {
+        if (i-- <= 0) {
+          v.push_back(form);
+        }
+      });
```

Comment:
I think this whole thing can be:
```cpp
auto buttons_form = dynamic_cast<GenericElement*>(in->at(0));
v.insert(v.end(), buttons_form.elts().begin(), buttons_form.elts_end());
```
the `apply_form` functions iterate over the entire tree, and we probably don't need that here - we just want to grab the enum names for `pad-buttons`

Reply:
It couldn't be that simple because that would return the form elements and not the forms themselves needed to make a new `GenericElement`, but this looks cleaner anyway.